op-batcher/batcher/driver.go

```diff
@@ -284,87 +285,129 @@ func (l *BatchSubmitter) calculateL2BlockRangeToStore(ctx context.Context) (eth.
 func (l *BatchSubmitter) loop() {
 	defer l.wg.Done()
 
-	ticker := time.NewTicker(l.PollInterval)
-	defer ticker.Stop()
+	loadTicker := time.NewTicker(l.PollInterval)
+	defer loadTicker.Stop()
+	publishTicker := time.NewTicker(100 * time.Millisecond)
+	defer publishTicker.Stop()
+	receiptsCh := make(chan txmgr.TxReceipt[txData])
+	queue := txmgr.NewQueue[txData](l.txMgr, l.MaxPendingTransactions, l.metr.RecordPendingTx)
+
 	for {
 		select {
-		case <-ticker.C:
+		case <-loadTicker.C:
 			l.loadBlocksIntoState(l.shutdownCtx)
-			l.publishStateToL1(l.killCtx)
+		case <-publishTicker.C:
+			_ = queue.TrySend(l.killCtx, l.publishStateToL1Factory(), receiptsCh)
+		case r := <-receiptsCh:
+			l.handleReceipt(r)
 		case <-l.shutdownCtx.Done():
-			l.publishStateToL1(l.killCtx)
+			l.drainState(receiptsCh, queue)
 			return
 		}
 	}
 }
 
-// publishStateToL1 loops through the block data loaded into `state` and
-// submits the associated data to the L1 in the form of channel frames.
-func (l *BatchSubmitter) publishStateToL1(ctx context.Context) {
-	for {
-		// Attempt to gracefully terminate the current channel, ensuring that no new frames will be
-		// produced. Any remaining frames must still be published to the L1 to prevent stalling.
-		select {
-		case <-ctx.Done():
-			err := l.state.Close()
-			if err != nil {
-				l.log.Error("error closing the channel manager", "err", err)
-			}
-		case <-l.shutdownCtx.Done():
-			err := l.state.Close()
-			if err != nil {
-				l.log.Error("error closing the channel manager", "err", err)
+func (l *BatchSubmitter) drainState(receiptsCh chan txmgr.TxReceipt[txData], queue *txmgr.Queue[txData]) {
+	err := l.state.Close()
+	if err != nil {
+		l.log.Error("error closing the channel manager", "err", err)
+	}
+
+	txDone := make(chan struct{})
+	go func() {
+		defer func() {
+			// wait for all transactions to complete
+			queue.Wait()
+			close(txDone)
+		}()
+
+		// keep publishing state until either:
+		//  - we've drained all pending data (EOF)
+		//  - an error occurs
+		//  - we get killed by the context
+		for {
+			select {
+			case <-l.killCtx.Done():
+				return
+			default:
+				err := queue.Send(l.killCtx, l.publishStateToL1Factory(), receiptsCh)
```

Comment:
Can you provide `l.publishStateToL1Job` instead of going through the factory?

Reply:
the reason for doing it through the factory is that `publishStateToL1Job` is called lazily only if there's room in the pending pool for the tx. If we called it directly, we would have a bunch of local state and nowhere to put it, including an `l1Tip` that may be out-of-date when the tx queue allows the tx later